### PR TITLE
Add focuspoint gravity to imgproxy url if focusArea is present in processing configuration

### PR DESCRIPTION
**Pull request by rtp-cgs**
Related Issue: https://github.com/somehow-digital/typo3-media-processing/issues/9

Hello, as mentioned in the issue a few weeks ago I prepared a pull request which passes down the focuspoint from the processing configuration to the imgproxy url.

Please let me know what you think about this and if you wan't me to do any optimizations. It would be great if this change could make it to your package.

Regards,
Cyrill

### CODE REVIEW

File: Classes/ImageService/ImgProxyImageService.php [hunk 1]
```diff
@@ -124,6 +126,14 @@ public function processTask(TaskInterface $task): ImageServiceResult
...
 		}
 
+		$focusArea = $configuration['focusArea'] ?? false;
+		if ($focusArea instanceof Area && !$focusArea->isEmpty()) {
+			$focusPointX = FocusAreaUtility::calculateCenter($focusArea->getOffsetLeft(), $focusArea->getWidth());
+			$focusPointY = FocusAreaUtility::calculateCenter($focusArea->getOffsetTop(), $focusArea->getHeight());
+
+			$uri->setGravity('fp', (float)$focusPointX, (float)$focusPointY);
+		}
+
```
Comment (thasmo) [inline Classes/ImageService/ImgProxyImageService.php:136]: Just to be in line with the other conditions, I'd prefer to have this not defining a separate variable. Also, I'm not sure that we need the type casting.

```suggestion
		if (isset($configuration['focusArea']) && !$configuration['focusArea']->isEmpty()) {
			$horizontalOffset = FocusAreaUtility::calculateCenter(
				$configuration['focusArea']->getOffsetLeft(),
				$configuration['focusArea']->getWidth(),
			);

			$verticalOffset = FocusAreaUtility::calculateCenter(
				$configuration['focusArea']->getOffsetTop(),
				$configuration['focusArea']->getHeight(),
			);

			$uri->setGravity('fp', $horizontalOffset, $verticalOffset);
		}

```

File: Classes/Utility/FocusAreaUtility.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+	 * @param \TYPO3\CMS\Core\Imaging\ImageManipulation\Area $area
+	 *
+	 * @return float|int
+	 */
+	public static function calculateCenter(float $position, float $size)
+	{
+		// The max method ensures that only positive floats are returned
+		return max(($position + ($size / 2)), 0);
+	}
+}
```
Comment (thasmo) [inline Classes/Utility/FocusAreaUtility.php:20]: Just to be consistent with the codebase, I'd prefer to not have doc blocks. Also, the comment is mostly superfluous due to the line being easy to read, I think.

```suggestion
class FocusAreaUtility
{
	public static function calculateCenter(float $offset, float $size)
	{
		return max(($offset + ($size / 2)), 0);
	}
}
```

File: Classes/UriBuilder/ImgProxyUri.php [hunk 3]
```diff
@@ -99,9 +99,9 @@ public function getType(): ?string
 		return $this->type;
 	}
 
-	public function setGravity(string $type, ?int $horizontalOffset, ?int $verticalOffset): self
+	public function setGravity(string $type, ?float $horizontalOffset, ?float $verticalOffset): self
 	{
-		$this->gravity = array_filter([$type, $horizontalOffset, $verticalOffset]);
+		$this->gravity = [$type, $horizontalOffset, $verticalOffset];
```
Comment (thasmo) [inline Classes/UriBuilder/ImgProxyUri.php:104]: Did you remove the array_filter for a reason? Can't remember exactly, but I think it fixed something; could be wrong, tho.

**Review by thasmo — CHANGES_REQUESTED** [review]
Thank you for writing the PR! :)

Have some feedback, nothing major tho.

File: Classes/UriBuilder/ImgProxyUri.php (hunk 3 above)
Comment (rtp-cgs) [reply Classes/UriBuilder/ImgProxyUri.php:104]: If the $horizontalOffset or $verticalOffset was 0 then it was filtered out, resulting in an invalid imgproxy url.

**Comment by rtp-cgs** [comment]
Hi, thanks for the fast response that is awesome.

I totally agree with your suggestions and just pushed them to the branch.

**Review by thasmo — APPROVED** [review]

**Comment by thasmo** [comment]
Thank you for your contribution!

Going to merge this soon, and I will also probably take a look at implementing the focal point/area feature for other services/libraries as well, if I find some time. I'd like to have this feature supported for all relevant services for v1.

**Pull request merged**